client/src/_session_layer.rs

```diff
@@ -0,0 +1,1536 @@
+                    e => return Err(e),
+                },
+            }
+        }
+
+        Err(SessionError::Generic(format!(
+            "All attempts to establish direct session with node [{node_id}] failed"
+        )))
+    }
+
+    async fn try_reverse_connection(
+        &self,
+        node_id: NodeId,
+        _permit: &SessionPermit,
+    ) -> Result<Arc<DirectSession>, SessionError> {
+        // We are trying to connect to Node without public IP. Send `ReverseConnection` message,
+        // so Node will connect to us.
+        if self.get_public_addr().await.is_none() {
+            return Err(SessionError::NotApplicable(
+                "We don't have public endpoints.".to_string(),
+            ));
+        }
+
+        log::info!(
+            "Request reverse connection. me={}, remote={node_id}",
+            self.config.node_id,
+        );
+
+        Err(SessionError::NotApplicable(
+            "ReverseConnection not implemented.".to_string(),
+        ))
+
+        //
+        // let mut awaiting = self.guarded.register_waiting_for_node(node_id).await?;
+        //
+        // let server_session = self.server_session().await?;
+        // server_session.reverse_connection(node_id).await?;
+        //
+        // log::debug!("ReverseConnection requested with node [{node_id}]");
+        //
+        // // We don't want to wait for connection finish, because we need longer timeout for this.
+        // // But if we won't receive any message from other Node, we would like to exit early,
+        // // to try out different connection methods.
+        // tokio::time::timeout(
+        //     self.config.reverse_connection_tmp_timeout,
+        //     awaiting.wait_for_first_message(),
+        // )
+        // .await?;
+        //
+        // // If we have first handshake message from other node, we can wait with
+        // // longer timeout now, because we can hope, that this node is responsive.
+        // let result = tokio::time::timeout(
+        //     self.config.reverse_connection_real_timeout,
+        //     awaiting.wait_for_connection(),
+        // )
+        // .await;
+        //
+        // match result {
+        //     Ok(Ok(session)) => {
+        //         log::info!("ReverseConnection - got session with node: [{node_id}]");
+        //         Ok(session)
+        //     }
+        //     Ok(Err(e)) => {
+        //         log::info!("ReverseConnection - failed to establish session with: [{node_id}]");
+        //         Err(e.into())
+        //     }
+        //     Err(_) => {
+        //         log::info!(
+        //             "ReverseConnection - waiting for session timed out ({}). Node: [{node_id}]",
+        //             humantime::format_duration(self.config.reverse_connection_real_timeout)
+        //         );
+        //         bail!("Not able to setup ReverseConnection within timeout with node: [{node_id}]")
+        //     }
+        // }
```

Comment:
Probably some todo / leftover

Reply:
Reverse connection PR is already implementing this: https://github.com/golemfactory/ya-relay/pull/250